src/build.c

```diff
@@ -2577,15 +2565,20 @@ void build_save_menu(GKeyFile *config, gpointer ptr, GeanyBuildSource src)
 				g_key_file_remove_key(config, build_grp_name, "error_regex", NULL);
 			if (pj->priv->build_filetypes_list != NULL)
 			{
-				data.config = config;
-				data.ft_names = g_ptr_array_new();
-				g_ptr_array_foreach(pj->priv->build_filetypes_list, foreach_project_filetype, (gpointer)(&data));
-				if (data.ft_names->pdata != NULL)
+				GPtrArray *ft_names = g_ptr_array_new();
+				guint i;
+				
+				foreach_ptr_array(ft, i, pj->priv->build_filetypes_list)
```

Comment:
IMO this would read better as:

```c
GPtrArray *build_fts = pj->priv->build_filetypes_list;
for (guint i = 0; i < build_fts->len; i++) {
    ft = build_fts->pdata[i];
    /// other code here...
}
```

so it doesn't hide code behind a macro. It's basically the same amount of code, and index iteration is something everyone already knows without thinking about it, unlike the macro that is named like a function.

Reply:
Agree with @codebrainz lower case local macros is BAD, please don't perpetuate the evil Geany already has.

Reply:
@codebrainz @elextr I patched in the pointer indexing code. We should add `@deprecated` to all the `foreach_*` macro doc-comments in the API, if they're not supposed to be used in new code, so people have a chance to find out.

Reply:
> lower case local macros is BAD

There are 6 of these in build.c, I can't work out who named them (it might need bisecting so obviously not worth it).

Reply:
> We should add @deprecated to all the foreach_* macro doc-comments in the API, if they're not supposed to be used in new code, so people have a chance to find out.

:+1: 

I don't think there's been any discussion about it (lately) but my personal opinion is that all macros and functions which obscure what the code does just to save a line or two of code should be deprecated. C is verbose, but it's easy to read if the code is all in sequence.

Reply:
@codebrainz @elextr 
BTW I wrote the macros not to 'save a line or two of code' but to encapsulate iteration so when seeing the macro invocation you know there are no bugs in iteration (that was the intent anyway).

I accept the counter-argument that things like array iteration are simple and a custom macro is a geany-specific thing to learn, and it could possibly make debugging more confusing if the macro was passed a wrongly typed parameter.

But what about things that people often get wrong, like not checking `doc->is_valid`, I think the `foreach_document` macro is worth having. `is_valid` *is* a Geany-specific thing, so it's easy to justify a macro here. Sure, make it upper-case though.

Reply:
@ntrel, agree there is some argument for the document loop one, since there isn't any really obvious solution to the `is_valid` thing.  (Aside, I thought the point of document IDs was to avoid dangling pointers, but my memory is a bit hazy and I don't have time just now to revise at this instant)

But if `FOREACH_DOCUMENT` is kept somehow it needs to be discoverable and documented, or people will only find it by accident reading existing code.  For example what does it do if the set of documents is changed by the user? Or is that just documented "don't do that"?

But for the others, I agree with @codebrainz that since most contributors to Geany don't work on it constantly, using standard C is most likely to be understood.

Reply:
> But what about things that people often get wrong, like not checking doc->is_valid, I think the foreach_document macro is worth having. is_valid is a Geany-specific thing, so it's easy to justify a macro here. Sure, make it upper-case though.

If there's an actual reason for it, sure. But what is the reason we keep invalid documents around? If there is a good reason to do that (like not premature optimization), I would tend to agree such a macro may be useful, but probably located in the `document.h` file and as we all agree upper-case.

Reply:
> it needs to be discoverable and documented, or people will only find it by accident reading existing code

Yes, that's why the `documents` macro has always mentioned it:

> * @see documents_array(). */
> #define documents ((GeanyDocument **)GEANY(documents_array)->pdata)

`documents_array` itself doesn't seem to be documented.

> For example what does it do if the set of documents is changed by the user?

It's undefined and not documented. I think it's to be expected that iterator invalidation is not handled by the macro.

Reply:
> what is the reason we keep invalid documents around? If there is a good reason to do that (like not premature optimization)

To avoid reading/writing to freed memory. The API allows people to read from a document pointer in a callback without seeing if they can find the document in documents_array. I don't think this can safely be deprecated, it would cause silent breakage.

> but probably located in the document.h file 

It already is.

Reply:
Moved to a new issue #2281